File: schema.jl
mutable struct Discriminator <: Comparable
    propertyName::String
    mapping::OrderedDict{String,String}
end

Discriminator() = Discriminator("", OrderedDict{String,String}())

function parse!(d::Discriminator, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "propertyName" s.propertyName = value end
        if key == "mapping"
            for (k,v) in value d.mapping[k] = v end
        end
    end
end

mutable struct XML <: Comparable
    name::String
    namespace::String
    prefix::String
    attribute::Bool
    wrapped::Bool
end

XML() = XML("", "", "", false, false)

function parse!(x::XML, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "name" s.name = value end
        if key == "namespace" s.namespace = value end
        if key == "prefix" s.prefix = value end
        if key == "attribute" s.attribute = value end
        if key == "wrapped" s.wrapped = value end
    end
end

mutable struct Schema <: Comparable
    discriminator::Discriminator
    xml::XML
    externalDocs::ExternalDocumentation
    ref::String
    type::String
    title::String
    summary::String
    description::String
    required::Vector{String}
    maximum::Union{Number,Nothing}
    exclusiveMaximum::Union{Number,Nothing}
    minimum::Union{Number,Nothing}
    exclusiveMinimum::Union{Number,Nothing}
    maxLength::Union{Number,Nothing}
    minLength::Union{Number,Nothing}
    pattern::String
    maxItems::Union{Number,Nothing}
    minItems::Union{Number,Nothing}
    format::String
    enum::Vector{String}
    allOf::Vector{Any}
    nullable::Union{Bool,Nothing}
    readOnly::Union{Bool,Nothing}
    writeOnly::Union{Bool,Nothing}
    deprecated::Union{Bool,Nothing}
    example::Any
    default::Any
    referenceURI::URI # To track external references.
    properties::OrderedDict{String,Schema}
end

Schema() = Schema(Discriminator(), XML(), ExternalDocumentation(), "", "", "", "", "",
    Vector{String}(), nothing, nothing, nothing, nothing, nothing, nothing, "", nothing, nothing,
    "", Vector{String}(), Vector{Any}(), nothing, nothing, nothing, nothing, "", "",
    URI(), OrderedDict{String,Schema}())

function parse!(s::Schema, data::OrderedDict{Any,Any})
    for (key,value) in data
        if key == "discriminator" parse!(s.discriminator, value) end
        if key == "xml" parse!(s.xml, value) end
        if key == "externalDocs" parse!(s.externalDocs, value) end
        if key == "\$ref" s.ref = value end
        if key == "type" s.type = value end
        if key == "title" s.title = value end
        if key == "summary" s.summary = value end
        if key == "description" s.description = value end
        if key == "required"
            for r in value push!(s.required, r) end
        end
        if key == "maximum" s.maximum = value end
        if key == "exclusiveMaximum" s.exclusiveMaximum = value end
        if key == "minimum" s.minimum = value end
        if key == "exclusiveMinimum" s.exclusiveMinimum = value end
        if key == "maxLength" s.maxLength = value end
        if key == "minLength" s.minLength = value end
        if key == "pattern" s.pattern = value end
        if key == "maxItems" s.maxItems = value end
        if key == "minItems" s.minItems = value end
        if key == "format" s.format = value end
        if key == "enum"
            for r in value push!(s.enum, r) end
        end
        if key == "nullable" s.nullable = value end
        if key == "readOnly" s.readOnly = value end
        if key == "writeOnly" s.writeOnly = value end
        if key == "deprecated" s.deprecated = value end
        if key == "allOf"
            s.allOf = value
            for v in value
                if v isa OrderedDict{Any,Any}
                    sc = Schema()
                    parse!(sc, v)
                    if isempty(s.ref) s.ref = sc.ref end
                    if isempty(s.type) s.type = sc.type end
                    if isempty(s.title) s.title = sc.title end
                    if isempty(s.summary) s.summary = sc.summary end
                    if isempty(s.description) s.description = sc.description end
                    if isempty(s.required) s.required = sc.required end
                    for (k,v) in sc.properties s.properties[k] = v end
                end
            end
        end
        if key == "properties" parse!(s.properties, value) end
    end
end

function parse!(s::OrderedDict{String,Schema}, data::OrderedDict{Any,Any})
    for (key,value) in data
        sc = Schema()
        parse!(sc, value)
        s[key] = sc
    end
end

File: src/parser/schema.cpp
//
// Created by Rakesh on 19/09/2025.
//

#include "parser.hpp"
#include "model/schema.hpp"

#include "model/example.hpp"

template <>
spt::model::Discriminator spt::parser::parse( c4::yml::ConstNodeRef node )
{
  auto m = model::Discriminator{};

  for ( const auto& child : node.children() )
  {
    if ( child.key() == "propertyName" ) child >> m.propertyName;
    if ( child.key() == "mapping" )
    {
      for ( const auto& map : child.children() )
      {
        if ( !map.has_val() ) continue;
        m.mapping.try_emplace( std::string{ std::string_view{ map.key() } }, std::string{ map.val().begin(), map.val().end() } );
      }
    }
  }

  return m;
}

template <>
spt::model::XML spt::parser::parse( c4::yml::ConstNodeRef node )
{
  auto m = model::XML{};

  for ( const auto& child : node.children() )
  {
    if ( child.key() == "name" ) child >> m.name;
    if ( child.key() == "namespace" ) child >> m._namespace;
    if ( child.key() == "prefix" ) child >> m.prefix;
    if ( child.key() == "attribute" ) child >> m.attribute;
    if ( child.key() == "wrapped" ) child >> m.wrapped;
  }

  return m;
}

namespace
{
  namespace pparser
  {
    using namespace spt;

    void parse( model::Schema& m, c4::yml::ConstNodeRef node )
    {
      for ( const auto& child : node.children() )
      {
        if ( child.key() == "$ref" ) child >> m.ref;
        if ( child.key() == "type" ) child >> m.type;
        if ( child.key() == "title" ) child >> m.title;
        if ( child.key() == "summary" ) child >> m.summary;
        if ( child.key() == "description" ) child >> m.description;
        if ( child.key() == "pattern" ) child >> m.pattern;
        if ( child.key() == "format" ) child >> m.format;
        if ( child.key() == "dialect" ) child >> m.dialect;
        if ( child.key() == "x-since-version" ) child >> m.sinceVersion;
        if ( child.key() == "nullable" ) child >> m.nullable;
        if ( child.key() == "readOnly" ) child >> m.readOnly;
        if ( child.key() == "writeOnly" ) child >> m.writeOnly;
        if ( child.key() == "deprecated" ) child >> m.deprecated;
        if ( child.key() == "discriminator" ) m.discriminator = parser::parse<model::Discriminator>( child );
        if ( child.key() == "xml" ) m.xml = parser::parse<model::XML>( child );
        if ( child.key() == "externalDocs" ) m.externalDocs = parser::parse<model::ExternalDocumentation>( child );
        if ( child.key() == "default" && child.has_val() ) m._default = std::string{ child.val().begin(), child.val().end() };
        if ( child.key() == "example" && child.has_val() ) m.example = std::string{ child.val().begin(), child.val().end() };
        if ( child.key() == "examples" )
        {
          for ( const auto& ex : child.children() )
          {
            if ( !ex.has_val() ) continue;
            m.examples.emplace_back( std::string{ ex.val().begin(), ex.val().end() } );
          }
        }

        if ( child.key() == "maximum" )
        {
          m.maximum.emplace();
          child >> *m.maximum;
        }

        if ( child.key() == "exclusiveMaximum" )
        {
          m.exclusiveMaximum.emplace();
          child >> *m.exclusiveMaximum;
        }

        if ( child.key() == "minimum" )
        {
          m.minimum.emplace();
          child >> *m.minimum;
        }

        if ( child.key() == "exclusiveMinimum" )
        {
          m.exclusiveMinimum.emplace();
          child >> *m.exclusiveMinimum;
        }

        if ( child.key() == "maxLength" )
        {
          m.maxLength.emplace();
          child >> *m.maxLength;
        }

        if ( child.key() == "minLength" )
        {
          m.minLength.emplace();
          child >> *m.minLength;
        }

        if ( child.key() == "maxItems" )
        {
          m.maxItems.emplace();
          child >> *m.maxItems;
        }

        if ( child.key() == "minItems" )
        {
          m.minItems.emplace();
          child >> *m.minItems;
        }

        if ( child.key() == "properties" )
        {
          for ( const auto& p : child.children() )
          {
            auto s = model::Schema{};
            parse( s, p );
            m.properties.try_emplace( std::string{ std::string_view{ p.key() } }, std::move( s ) );
          }
        }

        if ( child.key() == "items" )
        {
          m.items = std::make_unique<model::Schema>();
          parse( *m.items, child );
        }

        if ( child.key() == "allOf" )
        {
          m.allOf.reserve( 4 );
          for ( const auto& p : child.children() )
          {
            if ( p.is_map() ) m.allOf.emplace_back( spt::parser::parse<model::Schema>( p ) );
          }
        }

        if ( child.key() == "oneOf" )
        {
          m.oneOf.reserve( 4 );
          for ( const auto& p : child.children() )
          {
            if ( p.is_map() ) m.oneOf.emplace_back( spt::parser::parse<model::Schema>( p ) );
          }
        }

        if ( child.key() == "anyOf" )
        {
          m.anyOf.reserve( 4 );
          for ( const auto& p : child.children() )
          {
            if ( p.is_map() ) m.anyOf.emplace_back( spt::parser::parse<model::Schema>( p ) );
          }
        }
      }
    }
  }
}

template <>
void spt::parser::parse( model::Schema& m, c4::yml::ConstNodeRef node )
{
  pparser::parse( m, node );
}
